pkg/certificate/authorization/endpoint.go

```diff
@@ -0,0 +1,168 @@
+		authorization := existingGroup[0]
+		return &pb.AuthorizationResponse{
+			Token: authorization.Token.String(),
+		}, nil
+	}
+
+	createdGroup, err := endpoint.db.Create(ctx, req.UserId, 1)
+	if err != nil {
+		msg := "error creating authorization"
+		err := ErrEndpoint.New(msg)
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	groupLen := len(createdGroup)
+	if groupLen != 1 {
+		clientMsg := "error creating authorization"
+		internalMsg := clientMsg + fmt.Sprintf("; expected 1, got %d", groupLen)
+
+		endpoint.log.Error(internalMsg)
+		return nil, status.Error(codes.Internal, ErrEndpoint.New(clientMsg).Error())
+	}
+
+	authorization := createdGroup[0]
+
+	return &pb.AuthorizationResponse{
+		Token: authorization.Token.String(),
+	}, nil
+}
+
+// Run starts the endpoint HTTP server and waits for the context to be
+// cancelled or for `Close` to be called.
+func (endpoint *Endpoint) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	ctx, cancel := context.WithCancel(ctx)
+	var group errgroup.Group
+	group.Go(func() error {
+		<-ctx.Done()
+		return endpoint.server.Shutdown(context.Background())
+	})
+	group.Go(func() error {
+		defer cancel()
+		return endpoint.server.Serve(endpoint.listener)
+	})
+
+	return group.Wait()
+}
+
+// Close closes the endpoint HTTP server.
+func (endpoint *Endpoint) Close() error {
+	return endpoint.server.Close()
+}
+
+func (endpoint *Endpoint) httpCreate(writer http.ResponseWriter, httpReq *http.Request) {
+	var err error
+	ctx := context.Background()
+	mon.Task()(&ctx)(&err)
+
+	if httpReq.Method != http.MethodPost {
+		msg := fmt.Sprintf("unsupported HTTP method: %s", httpReq.Method)
+		err = ErrEndpoint.New(msg)
+		endpoint.log.Error(msg, zap.Error(ErrEndpoint.New(msg)))
+		http.Error(writer, msg, http.StatusBadRequest)
+		return
+	}
+
+	userID, err := ioutil.ReadAll(httpReq.Body)
+	if bytes.Equal([]byte{}, userID) {
+		msg := "missing user ID body"
+		err = ErrEndpoint.Wrap(err)
+		endpoint.log.Error(msg, zap.Error(err))
+		http.Error(writer, msg, http.StatusBadRequest)
+		return
+	}
+	if err != nil {
+		msg := "error reading body"
+		err = ErrEndpoint.Wrap(err)
+		endpoint.log.Error(msg, zap.Error(err))
+		http.Error(writer, msg, http.StatusInternalServerError)
+		return
+	}
+
+	authorizationReq := &pb.AuthorizationRequest{
+		UserId: string(userID),
+	}
+
+	authorizationRes, err := endpoint.Create(ctx, authorizationReq)
+	if err != nil {
+		msg := "error creating authorization"
+		err = ErrEndpoint.Wrap(err)
+		endpoint.log.Error(msg, zap.Error(err))
+		http.Error(writer, msg, http.StatusInternalServerError)
+		return
+	}
+
+	writer.WriteHeader(http.StatusCreated)
+	if _, err = writer.Write([]byte(authorizationRes.Token)); err != nil {
+		msg := "error writing response"
+		err = ErrEndpoint.Wrap(err)
+		endpoint.log.Error(msg, zap.Error(err))
+		http.Error(writer, msg, http.StatusInternalServerError)
```

Comment:
I'm not sure what the expected error code actually should be in this case because yes, there was an error but the resource was created.

Reply:
we cannot do anything in this case.
Just log and that's it.

The client won't get any response and if get's any, it will be the first status code

Reply:
I checked, the response body is the error message (instead of a token) and the status code is 201. I think this still accurately represents the state of the resource after the request and communicates the error message. So while we may not have any other choice at the moment, I think this is okay.

Reply:
Yes, that's correct.
It still applies that I've said in writing the header after the first write and that causes an error, however it won't be logged as I said because we use this helper function which ignores any error, hence, we don't get any error, we don't log anything.

Out of curiosity, how did you simulate a test which makes the `writer.Write` to return an error?

Reply:
I just changed the condition from `err != nil` to `err == nil` and ran the tests.